### PR TITLE
backend: limit number of accounts for bitbox01 to 1

### PR DESCRIPTION
**Pull request by benma**
In the app, we allow up to 5 accounts per coin for the BitBox02, and
one account per coin for BitBox01.

### CODE REVIEW

**Comment by x1ddos** [comment]
Just to double check: how does this work with account "explosion" for bb01, with different script types?

**Comment by benma** [comment]
> Just to double check: how does this work with account "explosion" for bb01, with different script types?

Excellent catch. In fact `len(accounts)` was not a good idea after all to get the next account number. Before it would not have worked with the BitBox01 (now disabled anyway).

I added a commit that gets the next account number by finding the maximum existing number.

File: backend/accounts.go [hunk 1]
```diff
@@ -257,12 +257,31 @@ func (backend *Backend) nextAccountNumber(coinCode coinpkg.Code, keystore keysto
 		return account.Configurations.ContainsRootFingerprint(rootFingerprint)
 	}
 
-	nextAccountNumber := uint16(len(backend.filterAccounts(accountsConfig, filter)))
+	nextAccountNumber := uint32(0)
```
Comment (x1ddos) [inline backend/accounts.go:260]: `int` would've worked just fine here. like I was saying, specific int types get in the way of readability. plus, you convert it to `uint16` in the end anyway. but, `AccoutNumber` returns `uint32`, unfortunately :(

so, you see, different conversions for not benefit.

File: backend/accounts.go [hunk 2]
```diff
@@ -257,12 +257,31 @@ func (backend *Backend) nextAccountNumber(coinCode coinpkg.Code, keystore keysto
...
+
+	if !keystore.SupportsMultipleAccounts() && nextAccountNumber >= 1 {
+		return 0, errp.WithStack(ErrAccountLimitReached)
+	}
 
 	if nextAccountNumber >= accountsHardLimit {
 		return 0, errp.WithStack(ErrAccountLimitReached)
 	}
-	return nextAccountNumber, nil
+	return uint16(nextAccountNumber), nil
```
Comment (x1ddos) [inline backend/accounts.go:284]: I know this cannot happen (famous last words), but this is a theoretical unchecked int overflow.

File: backend/accounts.go [hunk 3]
```diff
@@ -257,12 +257,31 @@ func (backend *Backend) nextAccountNumber(coinCode coinpkg.Code, keystore keysto
 		return account.Configurations.ContainsRootFingerprint(rootFingerprint)
 	}
 
-	nextAccountNumber := uint16(len(backend.filterAccounts(accountsConfig, filter)))
```
Comment (x1ddos) [inline backend/accounts.go:260]: To cover the case of bb01, would this work as well?

```go
if !keystore.SupportsMultipleAccounts() && len(backend.filterAccounts(accountsConfig, filter)) > 0 {
  return 0, ErrAccountLimitReached
}
```

Comment (benma) [reply backend/accounts.go:260]: yes but the code as it is now also covers it, right?

**Review by x1ddos — APPROVED** [review]

**Comment by benma** [comment]
@x1ddos added unit tests and slightly changed the function, pls check.

File: backend/keystore/keystore.go [hunk 4]
```diff
@@ -38,6 +38,7 @@ const (
 var ErrSigningAborted = errors.New("signing aborted by user")
 
 // Keystore supports hardened key derivation according to BIP32 and signing of transactions.
+//go:generate moq -pkg mocks -out mocks/keystore.go . Keystore
```
Comment (benma) [inline backend/keystore/keystore.go:41]: before you say anything @x1ddos, this tool is the one we looked at that produces mocks nearly like I'd write them by hand :D

File: backend/accounts_test.go [hunk 5]
```diff
@@ -21,11 +21,21 @@ import (
...
 	"github.com/digitalbitbox/bitbox-wallet-app/backend/signing"
+	"github.com/digitalbitbox/bitbox-wallet-app/util/errp"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
+func mustKeypath(keypath string) signing.AbsoluteKeypath {
+	kp, err := signing.NewAbsoluteKeypath(keypath)
+	if err != nil {
+		panic(err)
```
Comment (x1ddos) [inline backend/accounts_test.go:34]: A panic prevents all following tests from running, apart from avoiding an ugly stacktrace. A better way imho would be to fail only one specific test where this is used. for example,

```go
func makeKeypath(t *testing.T, p string) signing.AbsoluteKeypath {
  t.Helper()
  ...
  if err != nil {
    t.Fatalf("%s: %v", p, err)
  }
  return kp
}
```

File: backend/keystore/keystore.go (hunk 4 above)
Comment (x1ddos) [reply backend/keystore/keystore.go:41]: I remember :)
I still don't like the separate package `mocks`. It's often more useful in the same package as that one under test. plus, all those 10+ ugly sync.RWMutex.

**Review by x1ddos — APPROVED** [review]
LGTM

**Comment by benma** [comment]
> It's often more useful in the same package as that one under test. 

Such structs are often used in tests from multiple packages (any package that uses it and has tests). Where would you put it then? Example: BlockchainMock

**Comment by x1ddos** [comment]
> > It's often more useful in the same package as that one under test.
> 
> Such structs are often used in tests from multiple packages (any package that uses it and has tests). Where would you put it then? Example: BlockchainMock

If it is really used extensively in different package tests, maybe it deserves a dedicated testing package. Example: `net/http/httptest`. If I write `httptest.NewRequest(...)` it's very clear where it's coming from, compared to an anonymous `mock.Something`. Yeah sure, you can always rename the pkg at import but it's again an atypical use, like `int` vs `uint32`. Plus, If it's always called `mock` and you import multiple of them, you're really constrained to do named imports, picking random names.

**Pull request merged**